056-zk-kraft-migration.md

```diff
@@ -0,0 +1,435 @@
+The new `migration` value is added and used by the user to start the migration process which is automated by the operator taking the following actions:
+
+* deploy the controllers, configured with the ZooKeeper migration enabled, through the `KafkaNodePool` custom resources with the `role: controller` provided by the user.
+* roll the brokers, configured with the ZooKeeper migration enabled and connected to the newly created controllers.
+* run the actual metadata migration by the Apache Kafka migration component.
+
+Furthermore, the already available annotation values are using as following:
+
+* `enabled` to finalize the migration by disabling the ZooKeeper migration and removing the ZooKeeper connection information on controllers and brokers, rolling the nodes, and finally having a full KRaft-based cluster with ZooKeeper ensemble still running but not used by brokers anymore.
+* `disabled` to rollback the migration process, when still possible, to have brokers working again with the ZooKeeper ensemble only.
+
+Other "inputs" for the FSM are represented by:
+
+* the Kafka migration metrics checked by the operator during the actual metadata migration from ZooKeeper to KRaft.
+* the readiness of nodes (controllers or brokers) after deploying or rolling them, depending on the step in the process.
+
+The state machine is described in the following diagram:
+
+![Migration Finite State Machine](./images/056-zk-kraft-migration-fsm.png)
+
+The following sections explain the states and transitions in details.
+The "Phase"(s) reflect what is described in the official [ZooKeeper to KRaft Migration](https://kafka.apache.org/documentation/#kraft_zk_migration) guide in the Apache Kafka documentation but with the right degree of automation provided by the operator.
+Also, the `PRE_MIGRATION`, `MIGRATION` and `POST_MIGRATION` dotted boxes correspond to the `ZkMigrationState` [enum values](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/migration/ZkMigrationState.java#L32) from the Apache Kafka migration component.
+
+In order to make the process clearer as much as possible, each phase is described by:
+
+* _Preconditions_: conditions that should be met (i.e. current state in the migration FSM) in order to have the operator taking actions on a trigger.
+* _Trigger_: what makes the operator actions to run only with verified preconditions.
+* _Actions_: what the operator does when preconditions are met (i.e. nodes configuration, rolling, ...) together with things that could be observable during actions (i.e. Kafka metrics changes).
+* _Status changes_: how the FSM changes status when the actions are completed.
+
+#### Already running KRaft-based clusters
+
+**_Preconditions_**
+
+The Apache Kafka cluster is running in KRaft mode, so the `strimzi.io/kraft: enabled` annotation is already applied on the `Kafka` custom resource.
+The FSM is in the `KRaft` state (as per `Kafka.status.migrationState` field).
```

Comment:
Perhaps `metadataState`? What doesn't tie it so strongly to this being about the system in which metadata is stored, it's just telling you something about the metadata, which right now we're using for metadata migration states, but in the future we'd have a bit of flexibility. 

We could also perhaps change names of the intermediate states. I don't think they're very obvious to users right now. 

- ControllersInMigration -> EnablingKraftMigration
- Migration -> BuildingKraftMetadataLog
- DualWrite -> KraftDualWriting
- BrokersOutMigration -> FinalizingKraftMigration

I've inserted `Kraft` into each of these suggesions above on the assumption that we had decoupled the name of the field from migration. I'm also not overly attached to the _ing_ being in the names, but it does indicate that the state reflects something that is on-going.

Reply:
Yeah I should have changed the name already :-(
I will go with `metadataState`, which is pretty clear when no migration is happening with values like `ZooKeeper` (metadata are in ZK) and `KRaft` (metadata are in KRaft controllers) as well as when migration is running.

Regarding the other potential values, isn't it better to have the `KRaft` suffix for all of them (as for the dual write)?

* ControllersInMigration -> KRaftEnablingMigration
* Migration -> KRaftBuildingMetadataLog
* DualWrite -> KRaftDualWriting
* BrokersOutMigration -> KRaftFinalizingMigration

I think it would show immediately that's something related to KRaft.

Just to elaborate a little bit on this so we are on the same page ...

while the controllers are deployed, the state is still `ZooKeeper` which is anyway right because ... it reflects that metadata are still in ZooKeeper (while controllers are deployed by operator). Then we move to `KRaftEnablingMigration` which is more about "we have to enable the KRaft for brokers now, so rolling them".

the same would apply somehow on the finalization phase: while brokers are rolled, the state is still `KRaftDualWriting` which is right because ... it reflects that metadata are still written to both ZK and KRaft (while brokers are rolling). Then we move to `KRaftFinalizingMigration` which is more about "we have to remove controllers from using ZK, so rolling them".

Of course, it should clearly explained in the doc to provide the user a way to interpret the states.